spinnaker_graph_front_end/examples/hello_world_untimed/src/hello_world.c

```diff
@@ -0,0 +1,179 @@
+//! The data
+typedef struct char_data {
+    uint32_t max_index;
+    uint32_t n_chars;
+    uint8_t chars[];
+} char_data;
+static char_data *c_data;
+
+//! human readable definitions of each region in SDRAM
+typedef enum regions_e {
+    SYSTEM_REGION,
+    PARAMS_REGION,
+    RECORDED_DATA
+} regions_e;
+
+
+//! values for the priority for each callback
+typedef enum callback_priorities {
+    SDP = 0,
+    DMA = 1
+} callback_priorities;
+
+// -------------------------------------------------------------------
+
+static void record_data(void) {
+    bool recorded = recording_record(0, c_data->chars, c_data->n_chars * sizeof(char));
+    if (!recorded) {
+        log_error("Data was not recorded...");
+    }
+}
+
+
+//! \brief Initialises the recording parts of the model
+//! \return True if recording initialisation is successful, false otherwise
+static bool initialise_recording(void) {
+    data_specification_metadata_t *data = data_specification_get_data_address();
+    void *recording_region = data_specification_get_region(RECORDED_DATA, data);
+    bool success = recording_initialize(&recording_region, &recording_flags);
+    return success;
+}
+
+static void resume_callback(void);
+
+static void run(uint unused0, uint unused1) {
+    use(unused0);
+    use(unused1);
+
+    // Increment the step as it will start at "-1"
+    step++;
+
+    log_info("Running from step %u until %u or %u", step, n_steps, c_data->max_index);
+    while ((run_forever || step < n_steps) && step < c_data->max_index) {
+        if (recording_flags) {
+            record_data();
+            recording_do_timestep_update(step);
+        }
+        step++;
+    }
+
+    log_info("Steps run");
+    simulation_handle_pause_resume(resume_callback);
+    if (recording_flags) {
+        recording_finalise();
+    }
+    log_info("Making ready...");
+    simulation_ready_to_read();
+    log_info("Done");
+}
+
+// ! \brief Called when simulation is about to restart
+static void resume_callback(void) {
+    recording_reset();
+    log_info("Scheduling callback from resume");
+    spin1_schedule_callback(run, 0, 0, 1);
+}
+
+
+static void exit_callback() {
+    if (recording_flags) {
+        recording_finalise();
+    }
+}
+
+static void start_callback() {
+    log_info("Scheduling callback from start");
+    spin1_schedule_callback(run, 0, 0, 1);
+}
+
+static bool initialize() {
+    // Get the address this core's DTCM data starts at from SRAM
+    data_specification_metadata_t *data = data_specification_get_data_address();
+
+    // Read the header
+    if (!data_specification_read_header(data)) {
+        log_error("failed to read the data spec header");
+        return false;
+    }
+
+    // set up the simulation interface
+    uint32_t unused_timer_period;
+    if (!simulation_initialise(
+            data_specification_get_region(SYSTEM_REGION, data),
+            APPLICATION_NAME_HASH, &unused_timer_period, &n_steps,
+            &run_forever, &step, SDP, DMA)) {
```

Comment:
esp as step and run_forever seems not usful as well here.

Reply:
and n_steps

Reply:
run_forever and n_steps are both relevant here as explained elsewhere.